### PR TITLE
REQ-403 concurrency fixes part 1

### PR DESCRIPTION
**Pull request by lippirk**
This fixes parallel pool.ejects, and makes cert distribs a bit safer

Additional improvements (for a later PR):
- add a lock for the 'create bundle' script
- on startup, retrieve any missing certs from the master

### CODE REVIEW

**Comment by lippirk** [comment]
@lindig eventually the cert refresh code will need to be locked with this mutex. I don't mind adding it later, but worth keeping in mind

**Comment by psafont** [comment]
This looks reasonable to me, has this been tested?

**Review by psafont — APPROVED** [review]

**Comment by lippirk** [comment]
> This looks reasonable to me, has this been tested?

I've tested pool join, eject and verification enable manually

**Review by robhoes — APPROVED** [review]

File: ocaml/xapi/cert_distrib.ml [hunk 1]
```diff
@@ -345,6 +345,25 @@ let collect_pool_certs ~__context ~rpc ~session_id ~map ~from_hosts =
...
+let _distrib_m = Mutex.create ()
+
+(* where possible, the master host should control the certificate
+ * distributions.  this allows us to coordinate multiple parties that are
+ * trying to modify /etc/stunnel at the same time with [lock]!
+ *
+ * we apply this lock to all top level distribution calls, with the exception of
+ * the pool join functions that execute on the joiner.
+ *)
+let lock (f : unit -> 'a) : 'a =
```
Comment (lindig) [inline ocaml/xapi/cert_distrib.ml:357]: How about `with_lock` or `locked` as a name?

**Review by lindig — APPROVED** [review]

**Pull request merged**